packages/tpetra/core/src/Tpetra_Details_residual.hpp

```diff
@@ -249,63 +178,135 @@ void localResidual(const CrsMatrix<SC,LO,GO,NO> &  A,
-                lsum += A_val * X_lcl(A_row.colidx(iEntry),0);
-              }, A_x);
+              Kokkos::parallel_reduce(Kokkos::ThreadVectorRange (dev, row_length), [&] (const LO iEntry, residual_value_type& lsum) {
+                  const auto A_val = A_row.value(iEntry);
+                  lsum += A_val * X_colmap_lcl(A_row.colidx(iEntry),0);
+                }, A_x);
             
-            Kokkos::single(Kokkos::PerThread(dev),[&] () {            
-                R_lcl(lclRow,0) = B_lcl(lclRow,0) - A_x;
-              });           
-          });//end parallel_for TeamThreadRange
-      });//end parallel_for "residual-vector"
-  } else {
-    // MultiVector case
-    // Kernel interior shamelessly horked from Ifpack2_Details_ScaledDampedResidual_def.hpp
-    Kokkos::parallel_for("residual-multivector",policy,KOKKOS_LAMBDA(const team_member& dev) {
-        // NOTE: It looks like I should be able to get this data up above, but if I try to
-        // we get internal compiler errors.  Who knew that gcc tried to "gimplify"?
-        const LO numVectors = static_cast<LO>(X_lcl.extent(1));
-        Kokkos::parallel_for(Kokkos::TeamThreadRange (dev, 0, rows_per_team),[&] (const LO& loop) {
-            const LO lclRow = static_cast<LO> (dev.league_rank ()) * rows_per_team + loop;
+              Kokkos::single(Kokkos::PerThread(dev),[&] () {            
+                  R_lcl(lclRow,0) = B_lcl(lclRow,0) - A_x;
+                });           
+            });//end parallel_for TeamThreadRange
+        });//end parallel_for "residual-vector"
+    } else {
+      auto X_domainmap_lcl = X_domainmap->getLocalViewDevice(Access::ReadOnly);
+      
+      Kokkos::parallel_for("residual-vector",policy,KOKKOS_LAMBDA(const team_member& dev) {
+          Kokkos::parallel_for(Kokkos::TeamThreadRange (dev, 0, rows_per_team),[&] (const LO& loop) {
+              const LO lclRow = static_cast<LO> (dev.league_rank ()) * rows_per_team + loop;
+              const LO numRows = A_lcl.numRows ();
             
-            if (lclRow >= A_lcl.numRows ()) {
-              return;
-            }
-            const auto A_row = A_lcl.rowConst(lclRow);
-            const LO row_length = static_cast<LO> (A_row.length);
-            for(LO v=0; v<numVectors; v++) {
+              if (lclRow >= numRows) {
+                return;
+              }
+
+              const auto A_row = A_lcl.rowConst(lclRow);
+              const LO row_length = static_cast<LO> (A_row.length);
               residual_value_type A_x = KAT::zero ();          
-              
+            
               Kokkos::parallel_reduce(Kokkos::ThreadVectorRange (dev, row_length), [&] (const LO iEntry, residual_value_type& lsum) {
                   const auto A_val = A_row.value(iEntry);
-                  lsum += A_val * X_lcl(A_row.colidx(iEntry),v);
+                  const auto lclCol = A_row.colidx(iEntry);
+                  if (lclCol < numRows)
```

Comment:
That is *evil*

Reply:
I probably need to change the check to something like `lclCol < domainMap->getMaxLocalIndex()` or `lclCol < numRows+indexBase`.